File: Permission/MainActivityPermission.swift
// File: ios/Runner/AppDelegate.swift

import UIKit
import Flutter

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
    
    override func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    ) -> Bool {
        
        // Lấy flutter controller
        let controller : FlutterViewController = window?.rootViewController as! FlutterViewController
        
        // Tạo MethodChannel với tên trùng với Flutter và Android
        let permissionChannel = FlutterMethodChannel(name: "permission_settings",
                                                   binaryMessenger: controller.binaryMessenger)
        
        // Xử lý các method call từ Flutter
        permissionChannel.setMethodCallHandler({ [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            
            switch call.method {
            case "openAppSettings":
                // Mở trang cài đặt của ứng dụng
                self?.openAppSettings(result: result)
                
            case "openLocationSettings":
                // Mở trang cài đặt vị trí của hệ thống
                self?.openLocationSettings(result: result)
                
            case "openNotificationSettings":
                // Mở trang cài đặt thông báo của hệ thống
                self?.openNotificationSettings(result: result)
                
            default:
                result(FlutterMethodNotImplemented)
            }
        })
        
        GeneratedPluginRegistrant.register(with: self)
        return super.application(application, didFinishLaunchingWithOptions: launchOptions)
    }
    
    /**
     * Mở trang cài đặt của ứng dụng
     * Đây là nơi người dùng có thể cấp/thu hồi quyền cho ứng dụng
     */
    private func openAppSettings(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            // Kiểm tra xem URL cài đặt app có khả dụng không
            if let settingsUrl = URL(string: UIApplication.openSettingsURLString) {
                if UIApplication.shared.canOpenURL(settingsUrl) {
                    UIApplication.shared.open(settingsUrl) { success in
                        result(success)
                    }
                } else {
                    result(false)
                }
            } else {
                result(false)
            }
        }
    }
    
    /**
     * Mở trang cài đặt vị trí của hệ thống
     * Lưu ý: iOS không cho phép mở trực tiếp trang cài đặt vị trí
     * Chỉ có thể mở trang cài đặt chung
     */
    private func openLocationSettings(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            // iOS không cho phép mở trực tiếp trang Location Settings
            // Chỉ có thể mở Settings app hoặc app settings
            if let settingsUrl = URL(string: "App-Prefs:Privacy&path=LOCATION") {
                if UIApplication.shared.canOpenURL(settingsUrl) {
                    UIApplication.shared.open(settingsUrl) { success in
                        result(success)
                    }
                } else {
                    // Fallback: mở cài đặt app
                    self.openAppSettings(result: result)
                }
            } else {
                // Fallback: mở cài đặt app
                self.openAppSettings(result: result)
            }
        }
    }
    
    /**
     * Mở trang cài đặt thông báo
     * iOS không cho phép mở trực tiếp trang notification settings
     * Chỉ có thể mở trang cài đặt app
     */
    private func openNotificationSettings(result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            // iOS không cho phép mở trực tiếp trang Notification Settings
            // Chỉ có thể mở app settings
            self.openAppSettings(result: result)
        }
    }
}

/**
 * Extension để xử lý permissions (tùy chọn)
 */
extension AppDelegate {
    
    /**
     * Kiểm tra trạng thái quyền notification
     */
    func checkNotificationPermission(completion: @escaping (Bool) -> Void) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            DispatchQueue.main.async {
                completion(settings.authorizationStatus == .authorized)
            }
        }
    }
    
    /**
     * Yêu cầu quyền notification
     */
    func requestNotificationPermission(completion: @escaping (Bool) -> Void) {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, _ in
            DispatchQueue.main.async {
                completion(granted)
            }
        }
    }
}

File: Permission/MainActivityPermission.kt
// File: android/app/src/main/kotlin/com/yourpackage/MainActivity.kt

package com.yourpackage.yourapp

import android.content.Intent
import android.net.Uri
import android.provider.Settings
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

class MainActivity: FlutterActivity() {
    // Tên channel phải trùng với Flutter
    private val CHANNEL = "permission_settings"

    override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
        super.configureFlutterEngine(flutterEngine)
        
        // Tạo MethodChannel để giao tiếp với Flutter
        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL).setMethodCallHandler { call, result ->
            when (call.method) {
                "openAppSettings" -> {
                    // Mở trang cài đặt của ứng dụng
                    val success = openAppSettings()
                    result.success(success)
                }
                "openLocationSettings" -> {
                    // Mở trang cài đặt vị trí của hệ thống
                    val success = openLocationSettings()
                    result.success(success)
                }
                "openNotificationSettings" -> {
                    // Mở trang cài đặt thông báo của ứng dụng
                    val success = openNotificationSettings()
                    result.success(success)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }
    }

    /**
     * Mở trang cài đặt chi tiết của ứng dụng
     * Đây là nơi người dùng có thể cấp/thu hồi quyền
     */
    private fun openAppSettings(): Boolean {
        return try {
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                data = Uri.fromParts("package", packageName, null)
                // Thêm flags để đảm bảo intent hoạt động tốt
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
            }
            startActivity(intent)
            true
        } catch (e: Exception) {
            // Nếu không mở được trang cài đặt cụ thể, thử mở trang cài đặt chung
            try {
                val fallbackIntent = Intent(Settings.ACTION_SETTINGS)
                startActivity(fallbackIntent)
                true
            } catch (ex: Exception) {
                false
            }
        }
    }

    /**
     * Mở trang cài đặt vị trí của hệ thống
     */
    private fun openLocationSettings(): Boolean {
        return try {
            val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
            startActivity(intent)
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Mở trang cài đặt thông báo của ứng dụng (Android 8.0+)
     */
    private fun openNotificationSettings(): Boolean {
        return try {
            val intent = Intent().apply {
                when {
                    // Android 8.0 (API 26) trở lên
                    android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O -> {
                        action = Settings.ACTION_APP_NOTIFICATION_SETTINGS
                        putExtra(Settings.EXTRA_APP_PACKAGE, packageName)
                    }
                    // Android 5.0 (API 21) đến 7.1 (API 25)
                    android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP -> {
                        action = "android.settings.APP_NOTIFICATION_SETTINGS"
                        putExtra("app_package", packageName)
                        putExtra("app_uid", applicationInfo.uid)
                    }
                    // Phiên bản cũ hơn
                    else -> {
                        action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
                        data = Uri.fromParts("package", packageName, null)
                    }
                }
            }
            startActivity(intent)
            true
        } catch (e: Exception) {
            // Fallback: mở cài đặt app nếu không mở được cài đặt thông báo
            openAppSettings()
        }
    }
}

/**
 * Extension methods để xử lý permissions (tùy chọn)
 */
object PermissionHelper {
    
    /**
     * Kiểm tra xem có nên hiển thị rationale dialog không
     */
    fun shouldShowRequestPermissionRationale(activity: FlutterActivity, permission: String): Boolean {
        return activity.shouldShowRequestPermissionRationale(permission)
    }
    
    /**
     * Tạo intent để mở trang cài đặt cụ thể
     */
    fun createSettingsIntent(packageName: String, settingsAction: String = Settings.ACTION_APPLICATION_DETAILS_SETTINGS): Intent {
        return Intent(settingsAction).apply {
            data = Uri.fromParts("package", packageName, null)
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
    }
}